modules/contact/src/constraints/NormalNodalLMMechanicalContact.C

```diff
@@ -0,0 +1,204 @@
+
+  for (_j = 0; _j < _phi_master.size(); ++_j)
+    Ken(0, _j) += computeQpOffDiagJacobian(Moose::SlaveMaster, jvar);
+}
+
+Real NormalNodalLMMechanicalContact::computeQpResidual(Moose::ConstraintType /*type*/)
+{
+  std::map<dof_id_type, PenetrationInfo *>::iterator found =
+      _penetration_locator._penetration_info.find(_current_node->id());
+  if (found != _penetration_locator._penetration_info.end())
+  {
+    PenetrationInfo * pinfo = found->second;
+    if (pinfo != NULL)
+    {
+      Real a = -pinfo->_distance * _c;
+      Real b = _u_slave[_qp];
+
+      if (_ncp_type == "fb")
+        return a + b - std::sqrt(a * a + b * b + _epsilon);
+      else
+        return std::min(a, b);
+    }
+  }
+  return 0;
+}
+
+// Note that the Jacobians below are inexact. To really make them exact, the most algorithmically
+// rigorous way will be to accomplish libmesh/libmesh#2121
+
+Real NormalNodalLMMechanicalContact::computeQpJacobian(Moose::ConstraintJacobianType /*type*/)
+{
+  std::map<dof_id_type, PenetrationInfo *>::iterator found =
+      _penetration_locator._penetration_info.find(_current_node->id());
+  if (found != _penetration_locator._penetration_info.end())
+  {
+    PenetrationInfo * pinfo = found->second;
+    if (pinfo != NULL)
+    {
+      DualNumber<Real> dual_u_slave(_u_slave[_qp]);
+      dual_u_slave.derivatives() = 1.;
+
+      auto a = -pinfo->_distance * _c;
+      auto b = dual_u_slave;
+
+      if (_ncp_type == "fb")
+        return (a + b - std::sqrt(a * a + b * b + _epsilon)).derivatives();
+      else
+        return std::min(a, b).derivatives();
+    }
+  }
+  return 0;
+}
+
+Real
+NormalNodalLMMechanicalContact::computeQpOffDiagJacobian(Moose::ConstraintJacobianType type,
+                                                         unsigned jvar)
+{
+  std::map<dof_id_type, PenetrationInfo *>::iterator found =
+      _penetration_locator._penetration_info.find(_current_node->id());
+  if (found != _penetration_locator._penetration_info.end())
+  {
+    PenetrationInfo * pinfo = found->second;
+    if (pinfo != NULL)
+    {
+      DualNumber<Real> gap(-pinfo->_distance);
+
+      unsigned comp;
+      if (jvar == _master_var_num)
+        comp = 0;
+      else if (jvar == _disp_y_id)
+        comp = 1;
+      else if (jvar == _disp_z_id)
+        comp = 2;
+      else
+        return 0;
+
+      gap.derivatives() = pinfo->_normal(comp);
+
+      switch (type)
+      {
+        case Moose::SlaveSlave:
+          gap.derivatives() *= 1;
+          break;
+        case Moose::SlaveMaster:
+          gap.derivatives() *= -_phi_master[_j][_qp];
+          break;
+        default:
+          mooseError("LMs do not have a master contribution.");
+      }
+
+      auto a = gap * _c;
+      auto b = _u_slave[_qp];
+
+      if (_ncp_type == "fb")
+        return (a + b - std::sqrt(a * a + b * b + _epsilon)).derivatives();
+      else
+        return std::min(a, b).derivatives();
+    }
```

Comment:
`a` and `b` are always positive?

Reply:
The std::min function will ensure that either a or b will be zero, and consequently that the other will be positive